src/storage/mvcc/txn.rs

```diff
@@ -146,49 +146,63 @@ impl<S: Snapshot> MvccTxn<S> {
-                            || (write.write_type != WriteType::Delete
-                                && self.key_exist(&key, write.start_ts - 1)?)
-                        {
-                            return Err(Error::AlreadyExist {
-                                key: key.into_raw()?,
-                            });
-                        }
-                    }
-                }
+        if let Some(lock) = self.reader.load_lock(&key)? {
+            // Abort on lock belonging to other transaction.
+            if lock.ts != self.start_ts {
+                return Err(Error::KeyIsLocked {
+                    key: key.into_raw()?,
+                    primary: lock.primary,
+                    ts: lock.ts,
+                    ttl: lock.ttl,
+                });
+            }
+            if lock.lock_type != LockType::Pessimistic {
+                // Duplicated command. No need to overwrite the lock and data.
+                MVCC_DUPLICATE_CMD_COUNTER_VEC.prewrite.inc();
+                return Ok(());
             }
-            // ... or locks at any timestamp.
-            if let Some(lock) = self.reader.load_lock(&key)? {
-                if lock.ts != self.start_ts {
-                    return Err(Error::KeyIsLocked {
+        } else if options.prewrite_pessimistic_lock {
+            // Pessimistic lock does not exist, the transaction should be aborted.
+            info!("prewrite failed (pessimistic lock not found)";
+                    "start_ts" => self.start_ts,
+                    "key" => %key);
+
+            return Err(Error::PessimisticLockNotFound {
+                start_ts: self.start_ts,
+                key: key.into_raw()?,
+            });
+        }
+
+        if !options.skip_constraint_check {
+            if let Some((commit_ts, write)) = self.reader.seek_write(&key, u64::max_value())? {
+                // Abort on writes after our start timestamp ...
+                // If exists a commit version whose commit timestamp is larger than or equal to
+                // current start timestamp, we should abort current prewrite, even if the commit
+                // type is Rollback.
+                //
+                // If the transaction is pessimistic, the data conflict is resolved by `for_update_ts`
+                if commit_ts == self.start_ts
+                    || (options.is_pessimistic_lock.is_empty() && commit_ts > self.start_ts)
+                {
+                    MVCC_CONFLICT_COUNTER.prewrite_write_conflict.inc();
+                    return Err(Error::WriteConflict {
+                        start_ts: self.start_ts,
+                        conflict_start_ts: write.start_ts,
+                        conflict_commit_ts: commit_ts,
                         key: key.into_raw()?,
-                        primary: lock.primary,
-                        ts: lock.ts,
-                        ttl: lock.ttl,
+                        primary: primary.to_vec(),
```

Comment:
So for pessimistic lock, we just update the TTL in `prewrite`?

Reply:
The only thing changed is how it checks data conflict in pessimistic transactions. The normal procedure of `Prewrite` is not changed.

Reply:
One of the purposes of pessimistic transactions is that it must commit successfully if all the pessimistic locks are hold when prewrites.